File: backend/src/main/java/cz/cvut/kbss/ontodeside/patomat2/model/PatternInstance.java
package cz.cvut.kbss.ontodeside.patomat2.model;

import java.util.ArrayList;
import java.util.List;

public record PatternInstance(int id, String patternName, PatternMatch match, String sparqlInsert, String sparqlDelete,
                              List<NewEntity> newEntities) {

    public PatternInstance deepCopy() {
        return new PatternInstance(id, patternName, match, sparqlInsert, sparqlDelete, new ArrayList<>(newEntities.stream()
                                                                                                                .map(ne -> new NewEntity(ne.variableName(), ne.identifier(), ne.label()))
                                                                                                                .toList()));
    }
}

File: frontend/src/types/PatternInstance.ts
export type PatternInstance = {
  id: number;
  patternName: string;
  match: PatternMatch;
  sparqlInsert: string;
  sparqlDelete: string;
  newEntities: NewEntity[];
  status: boolean | null;
};

export type PatternMatch = {
  id: number;
  patternFile: string;
  bindings: ResultBinding[];
};

export type ResultBinding = {
  name: string;
  value: string;
  datatype: string;
  basedOnBlankNode?: boolean;
  label?: string;
};

export type ResultBindingInfo = ResultBinding & {
  bindingParts: ResultBindingParts;
  bindingColor: string;
};

export type UriParts = {
  base: string;
  localName: string;
};

export type ResultBindingParts = {
  prefix: string; // eg. "<"
  base: string; // eg. "http://cmt/"
  localName: string; // eg. "assignReviewer"
  suffix: string; // eg. ">" nebo "^^datatype"
  basedOnBlankNode?: boolean;
};

export type NewEntity = {
  identifier: string;
  variableName: string;
  labels: EntityLabel[];
};

export type EntityLabel = {
  value: string;
  property: string;
  apply: boolean;
};
